### PR TITLE
[CARBONDATA-3646] [CARBONDATA-3647]: Fix query failure with Index Server

### PR DESCRIPTION
**Pull request by vikramahuja1001**
**Problems:** 
1. Select * query  fails when using index server.
2. Filter query failure with Index server when loaded with global_sort_partition = 100000,  giving Null pointer exception

**Solution:**
1. Indexservertmp folder configuration has been changed. The Indexservertmp folder is now created outside the table, as it is not recommended to keep it inside the table.
2. table.getAbsoluteTableIdentifier()  was giving  NULL pointer exception when table was NULL, this case has been handled.  
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1299/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1320/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1308/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1374/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1395/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1378/

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala [hunk 1]
```diff
@@ -45,6 +47,8 @@ import org.apache.carbondata.processing.loading.events.LoadEvents.{LoadMetadataE
 import org.apache.carbondata.spark.rdd.SparkReadSupport
 import org.apache.carbondata.spark.readsupport.SparkRowReadSupportImpl
 
+
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: revert this change

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala [hunk 2]
```diff
@@ -125,6 +130,21 @@ class CarbonEnv {
     LOGGER.info("Initialize CarbonEnv completed...")
   }
 
+
+  def createIndexServerTmp(sparkSession: SparkSession): Unit = {
+    val conf = sparkSession.sparkContext.hadoopConfiguration
+    val fs = FileSystem.get(conf)
+    if (fs.getUri.toString.startsWith("hdfs://")) {
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: Use CarbonUtil.checkAndAppendFileSystemURIScheme to add URI to the path

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala (hunk 1 above)
Comment (vikramahuja1001) [reply integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala (hunk 2 above)
Comment (vikramahuja1001) [reply integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: done

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1388/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1400/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1384/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1379/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1389/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1401/

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala [hunk 3]
```diff
@@ -125,6 +127,19 @@ class CarbonEnv {
     LOGGER.info("Initialize CarbonEnv completed...")
   }
 
+
+  def createIndexServerTmp(sparkSession: SparkSession): Unit = {
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: We already have a method to create temp folder in CarbonUtil. Please use the same method and add the code to change the permission to 777

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 4]
```diff
@@ -3321,7 +3321,7 @@ public static String getIndexServerTempPath(String tablePath, String queryId) {
         .getProperty(CarbonCommonConstants.CARBON_INDEX_SERVER_TEMP_PATH);
     if (null == tempFolderPath) {
       tempFolderPath =
-          tablePath + "/" + CarbonCommonConstants.INDEX_SERVER_TEMP_FOLDER_NAME + "/" + queryId;
+          "/tmp/" + CarbonCommonConstants.INDEX_SERVER_TEMP_FOLDER_NAME + "/" + queryId;
```
Comment (kunal642) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: use CarbonCommonConstants.FILE_SEPARATOR instead on "/" in the method

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala [hunk 5]
```diff
@@ -71,6 +72,7 @@ class CarbonEnv {
       properties.addProperty(CarbonCommonConstants.STORE_LOCATION, storePath)
     }
     LOGGER.info(s"Initializing CarbonEnv, store location: $storePath")
+    createIndexServerTmp(sparkSession);
```
Comment (akashrn5) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: please add a comment, why is this required and what is this used for?

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala [hunk 6]
```diff
@@ -125,6 +127,19 @@ class CarbonEnv {
     LOGGER.info("Initialize CarbonEnv completed...")
   }
 
+
```
Comment (akashrn5) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: please add method level comments

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala [hunk 7]
```diff
@@ -125,6 +127,19 @@ class CarbonEnv {
...
+  def createIndexServerTmp(sparkSession: SparkSession): Unit = {
+    val path = CarbonUtil.checkAndAppendFileSystemURIScheme(
+      "/tmp/" + CarbonCommonConstants.INDEX_SERVER_TEMP_FOLDER_NAME)
+    if (!FileFactory.isFileExist(path)) {
+        // Create the new directory
+        LOGGER.info(s"Creating Index Server Tmp folder: ${path} ")
```
Comment (akashrn5) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: either give `tmp` or `temp` in log, remove Tmp

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonDropCacheCommand.scala [hunk 8]
```diff
@@ -49,18 +49,19 @@ case class CarbonDropCacheCommand(tableIdentifier: TableIdentifier, internalCall
     OperationListenerBus.getInstance.fireEvent(dropCacheEvent, operationContext)
 
     val cache = CacheProvider.getInstance().getCarbonCache
+    // Clearing cache from IndexServer
```
Comment (akashrn5) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonDropCacheCommand.scala]: change commnet from `clearing` to `clear`

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonDropCacheCommand.scala [hunk 9]
```diff
@@ -49,18 +49,19 @@ case class CarbonDropCacheCommand(tableIdentifier: TableIdentifier, internalCall
...
-        DataMapUtil.executeClearDataMapJob(carbonTable, DataMapUtil.DISTRIBUTED_JOB_NAME)
-      } else {
-        // Extract dictionary keys for the table and create cache keys from those
-        val dictKeys: List[String] = CacheUtil.getAllDictCacheKeys(carbonTable)
-        // Remove elements from cache
-        cache.removeAll(dictKeys.asJava)
-        DataMapStoreManager.getInstance().clearDataMaps(carbonTable.getAbsoluteTableIdentifier)
-      }
+      LOGGER.info("Clearing cache from driver side")
+      // Extract dictionary keys for the table and create cache keys from those
```
Comment (akashrn5) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonDropCacheCommand.scala]: please corect the comment, its not clear

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala (hunk 7 above)
Comment (vikramahuja1001) [reply integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala (hunk 6 above)
Comment (vikramahuja1001) [reply integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala (hunk 5 above)
Comment (vikramahuja1001) [reply integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: done

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 4 above)
Comment (vikramahuja1001) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala (hunk 3 above)
Comment (vikramahuja1001) [reply integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: done

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1404/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1394/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1416/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1396/

File: integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala [hunk 10]
```diff
@@ -71,6 +70,8 @@ class CarbonEnv {
       properties.addProperty(CarbonCommonConstants.STORE_LOCATION, storePath)
     }
     LOGGER.info(s"Initializing CarbonEnv, store location: $storePath")
+    // Creating the index server tepm folder where splits for select query is written
```
Comment (akashrn5) [inline integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: temp spelling is wrong

Comment (vikramahuja1001) [reply integration/spark2/src/main/scala/org/apache/spark/sql/CarbonEnv.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonDropCacheCommand.scala (hunk 9 above)
Comment (vikramahuja1001) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonDropCacheCommand.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonDropCacheCommand.scala (hunk 8 above)
Comment (vikramahuja1001) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonDropCacheCommand.scala]: done

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1405/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1400/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1417/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1409/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1423/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1410/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1401/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by MakKind** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1402/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1411/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1413/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1435/

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1421/

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by ydvpankaj97** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1417/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1425/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1437/

**Comment by akashrn5** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1452/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1569/

**Comment by vikramahuja1001** [comment]
@kunal642 , @akashrn5 , please give LGTM

**Comment by kunal642** [comment]
LGTM

**Comment by akashrn5** [comment]
LGTM

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1620/

**Comment by kunal642** [comment]
@vikramahuja1001 Please rebase

**Comment by vikramahuja1001** [comment]
@kunal642 , done

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1652/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1656/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1659/

**Comment by vikramahuja1001** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1668/

**Comment by vikramahuja1001** [comment]
@kunal642 , done

**Comment by kunal642** [comment]
LGTM

**Comment by akashrn5** [comment]
LGTM
